Common/Exceptions/PythonExceptionInterpreter.cs

```diff
@@ -0,0 +1,53 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using Python.Runtime;
+using QuantConnect.Util;
+
+namespace QuantConnect.Exceptions
+{
+    /// <summary>
+    /// Interprets <see cref="PythonExceptionInterpreter"/> instances
+    /// </summary>
+    public class PythonExceptionInterpreter : IExceptionInterpreter
+    {
+        /// <summary>
+        /// Determines the order that an instance of this class should be called
+        /// </summary>
+        public int Order => int.MaxValue;
+
+        /// <summary>
+        /// Determines if this interpreter should be applied to the specified exception. f
+        /// </summary>
+        /// <param name="exception">The exception to check</param>
+        /// <returns>True if the exception can be interpreted, false otherwise</returns>
+        public bool CanInterpret(Exception exception) => exception?.GetType() == typeof(PythonException);
+
+        /// <summary>
+        /// Interprets the specified exception into a new exception
+        /// </summary>
+        /// <param name="exception">The exception to be interpreted</param>
+        /// <param name="innerInterpreter">An interpreter that should be applied to the inner exception.
+        /// <returns>The interpreted exception</returns>
+        public Exception Interpret(Exception exception, IExceptionInterpreter innerInterpreter)
+        {
+            var pe = (PythonException)exception;
+
+            var message = pe.Message + PythonUtil.PythonExceptionStackParser(pe.StackTrace);
```

Comment:
I think we could move `PythonExceptionStackParser` from `PythonUtil` to this class since it is specific to python exception interpreters. What would be the best approach? A static method or a "normal" method? If the second, we could make all python exception interpreters inherit from `PythonExceptionInterpreter`.

Reply:
I kind of like it as a static method. Using inheritance solely for code reuse, in my opinion, is an anti-pattern since it limits the reuse to only sub-types. Leaving this as a static, stateless method makes it consumable anywhere and everywhere while keeping it easy to test.

Now, if we end up in a situation where we start adding options defining *how* a stack trace is parsed, I would recommend we put it into it's own class (preferably behind an interface to enable easier testing) and then use constructor dependency injection to give the implementation to all the python exception interpreters.